x-pack/plugins/apm/public/components/shared/FilterBar/ServiceEnvFilter.tsx

```diff
@@ -0,0 +1,94 @@
+import { EuiFilterGroup, EuiFilterButton, EuiSelect } from '@elastic/eui';
+import React, { useEffect, useRef, useState } from 'react';
+import { useFetcher } from '../../../hooks/useFetcher';
+import { loadServiceEnvironments } from '../../../services/rest/apm/services';
+import { useLocation } from '../../../hooks/useLocation';
+import { useUrlParams } from '../../../hooks/useUrlParams';
+import { history } from '../../../utils/history';
+import { fromQuery, toQuery } from '../Links/url_helpers';
+
+function without<T>(obj: T, key: string): T {
+  const result: { [objKey: string]: unknown } = Object.assign({}, obj);
+  delete result[key];
+  return result as T;
+}
+
+export const ServiceEnvFilter: React.FC = () => {
+  const location = useLocation();
+  const { urlParams } = useUrlParams();
+  const { start, end, kuery, environment } = urlParams;
+  const { data = [] } = useFetcher(
+    () => loadServiceEnvironments({ start, end, kuery, environment: '' }),
+    [start, end, kuery]
+  );
+  const availableEnvironments = ['Any', ...data, 'Not set'];
+  const [selectedValue, setSelectedValue] = useState(availableEnvironments[0]);
+  const isPostEffect = useRef(false);
```

Comment:
Why is a ref necessary?

Reply:
This whole component will likely change, but it was needed because the behavior of the initial render is different from later renders: It need to read from the URL to select the correct option once the available environments load and populate the drop down. after that initial render it will go the other direction and update the URL based on the selection.

Reply:
> it was needed because the behavior of the initial render is different from later renders: It need to read from the URL to select the correct option once the available environments load and populate the drop down. after that initial render it will go the other direction and update the URL based on the selection.

I think whenever possible we should avoid conditional state. Eg. Either it should always read environment from props or it should never (and instead use own state).
In this case why can't we always read from the url?